src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+    }
+
+    static const std::array<nc_color, 7> colors {{
+        c_green,
+        c_ltgreen,
+        c_yellow,
+        c_magenta,
+        c_ltred,
+        c_red,
+        c_ltgray
+    }};
+
+    bar_color = colors[(int) roundf( colors.size() * percentage )];
+}
+
+void health_bar::set_health_percentage( float percentage )
+{
+    bool overloaded = false;
+    if( percentage > 1 ) {
+        current_health = max_health;
+        overloaded = true;
+    } else if( percentage < 0 ) {
+        current_health = 0;
+    } else {
+        current_health = percentage * max_health;
+    }
+
+    refresh_bar( overloaded, percentage );
+}
+
+smiley_indicator::smiley_indicator( int x, int y, ui_anchor anchor) : ui_element(2, 1, x, y, anchor )
+{
+    set_state( neutral );
+}
+
+void smiley_indicator::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), smiley_color, smiley_str.c_str() );
+}
+
+void smiley_indicator::set_state( smiley_state new_state )
+{
+    state = new_state;
+    switch( new_state ) {
+        case very_unhappy:
+            smiley_color = c_red;
+            smiley_str = "D:";
+            break;
+        case unhappy:
+            smiley_color = c_ltred;
+            smiley_str = ":(";
+            break;
+        case neutral:
+            smiley_color = c_white;
+            smiley_str = ":|";
+            break;
+        case happy:
+            smiley_color = c_ltgreen;
+            smiley_str = ":)";
+            break;
+        case very_happy:
+            smiley_color = c_green;
+            smiley_str = ":D";
+            break;
+    }
+}
+
+template<class T>
+tile_panel<T>::tile_panel( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor )
+                       : ui_element( size_x, size_y, x, y, anchor )
+{
+    num_tiles = size_x * size_y;
+    tiles = new T[num_tiles];
+}
+
+template<class T>
+void tile_panel<T>::set_rect( const ui_rect &new_rect )
+{
+    ui_element::set_rect( new_rect );
+    num_tiles = new_rect.size_x * new_rect.size_y;
+    delete[] tiles;
+    tiles = new T[num_tiles];
+}
+
+template<class T>
+tile_panel<T>::~tile_panel()
+{
+    delete[] tiles;
+}
+
+template<class T>
+void tile_panel<T>::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    for( unsigned int x = 0; x < get_rect().size_x; x++ ) {
+        for( unsigned int y = 0; y < get_rect().size_y; y++ ) {
+            tiles[y * get_rect().size_x + x].draw( win, x + get_ax(), y + get_ay() );
+        }
+    }
+}
+
+template<class T>
+void tile_panel<T>::set_tile( const T &tile, unsigned int x, unsigned int y )
+{
+    if( x >= get_rect().size_x || y >= get_rect().size_y ) {
+        return; // TODO: give feedback
+    }
+
+    int index = x * get_rect().size_y + y; // TODO: why does this need to be reversed?
+
+    tiles[index] = tile; // Does this call T's copy constructor? or maybe of a derived type?
```

Comment:
> Does this call T's copy constructor?

Yes, it does.

> or maybe of a derived type?

Nope. It calls `T::operator=(const T&)`, because those are the involved types. But if `tile` is actually a reference to a derived type, it will most likely be sliced as `T::operator=` wont recognize it.